### PR TITLE
ORCA: allow not enforce distribution key in 3-stage aggregate

### LINKED ISSUE #659: [Bug] Optimizer could choose HashAggregate to dedup based on cost.
### Cloudberry Database version

main

### What happened

Optimizer could produce better plan to dedup based on cost.

### What you think should happen instead

_No response_

### How to reproduce

For the planner, we need to create better plan which uses HashAggregate to dedup columns.
For example,
```
ebi_dwh=# explain analyze WITH UniqueIDs AS (
    SELECT DISTINCT gsp_id
    FROM eft_com_call_statistic
)
SELECT COUNT(*)
FROM UniqueIDs;
                                                                       QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=0.00..1419.47 rows=1 width=8) (actual time=2510.079..2510.081 rows=1 loops=1)
   ->  Gather Motion 4:1  (slice1; segments: 4)  (cost=0.00..1419.47 rows=1 width=8) (actual time=1998.786..2510.054 rows=4 loops=1)
         ->  Partial Aggregate  (cost=0.00..1419.47 rows=1 width=8) (actual time=1993.802..1993.803 rows=1 loops=1)
               ->  HashAggregate  (cost=0.00..1419.47 rows=2315005 width=1) (actual time=1294.980..1855.013 rows=2339996 loops=1)
                     Group Key: gsp_id
                     ->  Seq Scan on eft_com_call_statistic  (cost=0.00..969.33 rows=2375680 width=37) (actual time=0.343..339.551 rows=2377251 loops=1)
 Planning Time: 5.357 ms
   (slice0)    Executor memory: 98354K bytes.
   (slice1)    Executor memory: 244665K bytes avg x 4x(0) workers, 244755K bytes max (seg2).  Work_mem: 286737K bytes max.
 Memory used:  2097152kB
 Optimizer: Pivotal Optimizer (GPORCA)
 Execution Time: 2584.618 ms
(12 rows)

```

This plan uses HashAggregate to perform gsp_id deduplication, while

```
 ebi_dwh=# explain analyze select count(distinct gsp_id) from eft_com_call_statistic;
                                                                    QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=0.00..1142.67 rows=1 width=8) (actual time=10573.416..10573.418 rows=1 loops=1)
   ->  Gather Motion 4:1  (slice1; segments: 4)  (cost=0.00..1142.67 rows=1 width=8) (actual time=10205.645..10573.387 rows=4 loops=1)
         ->  Partial Aggregate  (cost=0.00..1142.67 rows=1 width=8) (actual time=10437.005..10437.007 rows=1 loops=1)
               ->  Seq Scan on eft_com_call_statistic  (cost=0.00..969.33 rows=2375680 width=37) (actual time=0.269..548.673 rows=2377251 loops=1)
 Planning Time: 4.608 ms
   (slice0)    Executor memory: 37K bytes.
   (slice1)    Executor memory: 148480K bytes avg x 4x(0) workers, 148579K bytes max (seg2).
 Memory used:  2097152kB
 Optimizer: Pivotal Optimizer (GPORCA)
 Execution Time: 10573.853 ms
(10 rows)
```

It seems the cost estimation is no accurate.

### Operating System

No specific

### Anything else

No

### Are you willing to submit PR?

- [ ] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://cloudberrydb.org/community/coc).

### PR DESCRIPTION
**Pull request by jiaqizho**
Fixes #659

### Change log

In the case where the group-by key is a distribution key, ORCA disables 3-stage aggregate by default.

Add a new GUC(optimizer_enable_use_distribution_in_dqa) to allow not enforce distribution key in 3-stage aggregate in ORCA. which enables local deduplication before the 2-stage agg.

### Type of Change
- [ ] Bug fix (non-breaking change)
- [ ] New feature (non-breaking change)
- [ ] Breaking change (fix or feature with breaking changes)
- [ ] Documentation update

### Test Plan

- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] Passed `make installcheck`
- [ ] Passed `make -C src/test installcheck-cbdb-parallel`

### Impact

**Performance:**

**User-facing changes:**

**Dependencies:**

### Checklist
- [ ] Followed [contribution guide](https://cloudberry.apache.org/contribute/code)
- [ ] Added/updated documentation
- [ ] Reviewed code for security implications
- [ ] Requested review from [cloudberry committers](https://github.com/orgs/apache/teams/cloudberry-committers)

### CI Skip Instructions

---

### CODE REVIEW

**Comment by jiaqizho** [comment]
The context about this PR: ORCA can create the 3-stage agg in `agg(distinct)` path, but prohibit the distribution key in distinct. Because ORCA think 2-stage agg is enough.

So u can see the 3-stage agg path but no cost generated in the memo. like 
```
ROOT
Group 9 (#GExprs: 8):
  0: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]  [ 0 8 ]
  1: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Three Stage Scalar DQA  ]  [ 12 15 ]
  2: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Two Stage Scalar DQA  ]  [ 16 15 ]
  3: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Two Stage Scalar DQA  ]  [ 20 25 ]
  4: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Two Stage Scalar DQA ]  [ 20 25 ]
    Cost Ctxts:
      main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 484.101386
  5: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Two Stage Scalar DQA ]  [ 16 15 ]
    Cost Ctxts:
      main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 679.103561
  6: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Three Stage Scalar DQA ]  [ 12 15 ]
    Cost Ctxts:
  7: CPhysicalScalarAgg( Global ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ]  [ 0 8 ]
    Cost Ctxts:
      main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 527.493765
  Grp OptCtxts:
    0 (stage 0): (req CTEs: [], req order: [<empty> match: satisfy ], req dist: [SINGLETON (master) match: satisfy], req rewind: [], req rewind: [NONE NO-MOTION match: satisfy], req partition propagation: [<empty> match: satisfy ]) => Best Expr:4
```

The GroupExpression#6 is the **Three Stage Scalar DQA**.

Details of why this path have no cost:
1. ORCA will create a `CPhysicalMotionHashDistribute HASHED` GroupExpression(Redistribute Motion) below the GroupExpression#6. If current physical agg enforce the distribution.
2. If current required distribution key(hashagg) is same as distribution key, then `Redistribute Motion` won't get the cost(in CPhysicalMotion::FValidContext)
3. So the root group(GroupExpression#6) won't be selected.

**Comment by my-ship-it** [comment]
Just tried, but cannot produce 3-stage aggregate still, the cost is still higher than 2-stage aggregate.
```
Group 9 (#GExprs: 8):
  0: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]  [ 0 8 ]
  1: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Three Stage Scalar DQA  ]  [ 12 15 ]
  2: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Two Stage Scalar DQA  ]  [ 16 15 ]
  3: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Two Stage Scalar DQA  ]  [ 20 25 ]
  4: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Two Stage Scalar DQA ]  [ 20 25 ]
    Cost Ctxts:
      main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 566.388669
  5: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Two Stage Scalar DQA ]  [ 16 15 ]
    Cost Ctxts:
      main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 1090.678248
  6: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Three Stage Scalar DQA ]  [ 12 15 ]
    Cost Ctxts:
  7: CPhysicalScalarAgg( Global ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ]  [ 0 8 ]
    Cost Ctxts:
      main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 765.687983
  Grp OptCtxts:
    0 (stage 0): (req CTEs: [], req order: [<empty> match: satisfy ], req dist: [SINGLETON (master) match: satisfy], req rewind: [], req rewind: [NONE NO-MOTION match: satisfy], req partition propagation: [<empty> match: satisfy ]) => Best Expr:4
```

**Comment by jiaqizho** [comment]
> [ ](/jiaqizho)

also need running with GUC `optimizer_force_three_stage_scalar_dqa`/`optimizer_force_multistage_agg`.

updated the test cases...

**Comment by jiaqizho** [comment]
> Just tried, but cannot produce 3-stage aggregate still, the cost is still higher than 2-stage aggregate.
> 
> ```
> Group 9 (#GExprs: 8):
>   0: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]  [ 0 8 ]
>   1: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Three Stage Scalar DQA  ]  [ 12 15 ]
>   2: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Two Stage Scalar DQA  ]  [ 16 15 ]
>   3: CLogicalGbAgg( Global ) Grp Cols: [][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ] , m_aggStage :[  Two Stage Scalar DQA  ]  [ 20 25 ]
>   4: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Two Stage Scalar DQA ]  [ 20 25 ]
>     Cost Ctxts:
>       main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 566.388669
>   5: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Two Stage Scalar DQA ]  [ 16 15 ]
>     Cost Ctxts:
>       main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 1090.678248
>   6: CPhysicalScalarAgg( Global, multi-stage ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ] , m_aggStage :[ Three Stage Scalar DQA ]  [ 12 15 ]
>     Cost Ctxts:
>   7: CPhysicalScalarAgg( Global ) Grp Cols: [], Minimal Grp Cols:[], Generates Duplicates :[ 0 ]  [ 0 8 ]
>     Cost Ctxts:
>       main ctxt (stage 0)0.1, child ctxts:[0], rows:1.000000 (group), cost: 765.687983
>   Grp OptCtxts:
>     0 (stage 0): (req CTEs: [], req order: [<empty> match: satisfy ], req dist: [SINGLETON (master) match: satisfy], req rewind: [], req rewind: [NONE NO-MOTION match: satisfy], req partition propagation: [<empty> match: satisfy ]) => Best Expr:4
> ```


also need running with GUC `optimizer_force_three_stage_scalar_dqa`/`optimizer_force_multistage_agg`.

updated the test cases...

**Review by my-ship-it — APPROVED** [review]
Good work. LGTM

**Review by tuhaihe — DISMISSED** [review]

File: src/backend/gpopt/config/CConfigParamMapping.cpp [hunk 1]
```diff
@@ -290,7 +290,11 @@ CConfigParamMapping::SConfigMappingElem CConfigParamMapping::m_elements[] = {
 	 false,	 // m_negate_param
 	 GPOS_WSZ_LIT(
 		 "Explore a nested loop join even if a hash join is possible")},
-
+	{EopttraceEnableUseDistributionInDQA,
+	 &optimizer_enable_use_distributetion_in_dqa,
```
Comment (jiaqizho) [inline src/backend/gpopt/config/CConfigParamMapping.cpp:294]: distributetion -> distribution

**Review by my-ship-it — APPROVED** [review]

**Review by tuhaihe — APPROVED** [review]

**Pull request merged**